commit bf25a08cc58c8e872806e92dd36d9ff91e5523d9
Author: Laura Flores <lflores@ibm.com>
Date:   Tue Jul 29 22:46:46 2025 +0000

    mgr, common, qa, doc: issue health error after max expiration is exceeded
    
    ----------------- Enhancement to the Original Fix -----------------
    
    During a mgr failover, the active mgr is marked available if:
      1. The mon has chosen a standby to be active
      2. The chosen active mgr has all of its modules initialized
    
    Now that we've improved the criteria for sending the "active" beacon
    by enforcing it to retry initializing mgr modules, we need to account
    for extreme cases in which the modules are stuck loading for a very long
    time, or even indefinitely. In these extreme cases where the modules might
    never initialize, we don't want to delay sending the "active" beacon for
    too long. This can result in blocking other important mgr functionality,
    such as reporting PG availability in the health status. We want
    to avoid sending warnings about PGs being unknown in the health status when
    that's not ultimately the problem.
    
    To account for an exeptionally long module loading time, I added a new
    configurable `mgr_module_load_expiration`. If we exceed this maximum amount
    of time (in ms) allotted for the active mgr to load the mgr modules before declaring
    availability, the standby will then proceed to mark itself "available" and
    send the "active" beacon to the mon and unblock other critical mgr functionality.
    
    If this happens, a health error will be issued at this time, indicating
    which mgr modules got stuck initializing (See src/mgr/PyModuleRegistry.cc). The
    idea is to unblock the rest of the mgr's critical functionality while making it
    clear to Ceph operators that some modules are unusable.
    
    --------------------- Integration Testing --------------------
    
    The workunit was rewritten so it tests for these scenarios:
    
    1. Normal module loading behavior (no health error should be issued)
    2. Acceptable delay in module loading behavior (no health error should be
       issued)
    3. Unacceptable delay in module loading behavior (a health error should be
       issued)
    
    --------------------- Documentation --------------------
    
    This documentation explains the "Module failed to initialize"
    cluster error.
    
    Users are advised to try failing over
    the mgr to reboot the module initialization process,
    then if the error persists, file a bug report. I decided
    to write it this way instead of providing more complex
    debugging tips such as advising to disable some mgr modules
    since every case will be different depending on which modules
    failed to initialize.
    
    In the bug report, developers can ask for the health detail
    output to narrow down which module is causing a bottleneck,
    and then ask the user to try disabling certain modules until
    the mgr is able to fully initialize.
    
    Fixes: https://tracker.ceph.com/issues/71631
    Signed-off-by: Laura Flores <lflores@ibm.com>

diff --git a/src/mgr/Mgr.cc b/src/mgr/Mgr.cc
index b48a9262f21..2257a8ca718 100644
--- a/src/mgr/Mgr.cc
+++ b/src/mgr/Mgr.cc
@@ -70,7 +70,8 @@ Mgr::Mgr(MonClient *monc_, const MgrMap& mgrmap,
   clog(clog_),
   audit_clog(audit_clog_),
   initialized(false),
-  initializing(false)
+  initializing(false),
+  initialization_start_time(ceph::coarse_mono_clock::zero())
 {
   cluster_state.set_objecter(objecter);
 }
@@ -168,6 +169,7 @@ void Mgr::background_init(Context *completion)
   ceph_assert(!initializing);
   ceph_assert(!initialized);
   initializing = true;
+  initialization_start_time = ceph::coarse_mono_clock::now();
 
   finisher.start();
 
@@ -770,6 +772,32 @@ bool Mgr::got_mgr_map(const MgrMap& m)
   return false;
 }
 
+bool Mgr::exceeded_initialization_expiration()
+{
+  // initialization_start_time=0 when initialization hasn't started yet,
+  // so know we can't have exceeded the time expiration.
+  if (ceph::coarse_mono_clock::is_zero(initialization_start_time)) {
+    return false;
+  }
+
+  // Save the amount of time elapsed
+  auto time_elapsed = ceph::coarse_mono_clock::now() - initialization_start_time;
+  dout(20) << "time elapsed since mgr initialization: " << time_elapsed << dendl;
+
+  // Reset start time if the expiration time has been exceeded.
+  // Signal initialization=true so the mgr forcibly sends an "active" beacon
+  auto expiration = g_conf().get_val<std::chrono::milliseconds>("mgr_module_load_expiration");
+  bool exceeded_expiration = time_elapsed > expiration;
+  if (exceeded_expiration) {
+    std::lock_guard l(lock);
+    initialization_start_time = ceph::coarse_mono_clock::zero();
+    initializing = false;
+    initialized = true;
+  }
+
+  return exceeded_expiration;
+}
+
 void Mgr::handle_mgr_digest(ref_t<MMgrDigest> m)
 {
   dout(10) << m->mon_status_json.length() << dendl;
diff --git a/src/mgr/Mgr.h b/src/mgr/Mgr.h
index 638f246d73f..ec3bad9f42b 100644
--- a/src/mgr/Mgr.h
+++ b/src/mgr/Mgr.h
@@ -67,6 +67,7 @@ protected:
 
   bool initialized;
   bool initializing;
+  ceph::coarse_mono_time initialization_start_time;
 
 public:
   Mgr(MonClient *monc_, const MgrMap& mgrmap,
@@ -76,6 +77,7 @@ public:
   ~Mgr();
 
   bool is_initialized() const {return initialized;}
+  bool exceeded_initialization_expiration();
   entity_addrvec_t get_server_addrs() const {
     return server.get_myaddrs();
   }
